component/src/main/java/io/siddhi/extension/execution/string/FillTemplateFunctionExtension.java

```diff
@@ -118,23 +124,45 @@
     @Override
     protected Object execute(Object[] objects, State state) {
         String sourceString = (String) objects[0];
+        Map<String, Object> valueMap = null;
         Matcher templateMatcher = templatePattern.matcher(sourceString);
         String match;
         Matcher indexMatcher;
-        int index;
-        while (templateMatcher.find()) {
-            match = templateMatcher.group(0);
-            indexMatcher = indexPattern.matcher(match);
-            if (indexMatcher.find()) {
-                index = Integer.parseInt(indexMatcher.group(0));
-                if (index < 1 || index >= objects.length) {
-                    throw new SiddhiAppRuntimeException("Index given for template elements "
-                            + "should be greater than 0 and less than '" + objects.length + "'. But found "
-                            + index + " in" + " the template '" + sourceString + "'.");
+        if (objects[1] instanceof HashMap) {
+            valueMap = (HashMap<String, Object>) objects[1];
+            String key = "";
+            while (templateMatcher.find()) {
+                match = templateMatcher.group(0);
+                indexMatcher = indexPattern.matcher(match);
+                if (indexMatcher.find()) {
+                    key = indexMatcher.group(0);
+                    if (key == null || key.equals("")) {
+                        throw new SiddhiAppRuntimeException("Key given for template elements "
+                                + "should be greater not null or empty. But found "
+                                + " in" + " the template '" + sourceString + "'.");
+                    }
+                    if (valueMap.get(key) != null) {
+                        sourceString = sourceString.replace(match, valueMap.get(key).toString());
+                    }
+                }
+            }
+        } else {
+            int index;
+            while (templateMatcher.find()) {
```

Comment:
templateMatcher.find() is common for both if/else, take it out

Reply:
I think we can't do that since the logic of the while block different from each case.

Reply:
don't use regex template mapper, it is costly, try to use simple StringBuilder, also if the template is a constant try to do some pre-processing (such as splitting the strings by `{{`or`}}` and getting the list of both text and key list ) so that you can rebuild faster at runtime.